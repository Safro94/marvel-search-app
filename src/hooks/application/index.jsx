import React, { createContext, useContext, useReducer } from 'react';

const SET_THEME = 'SET_THEME';
const SET_CHARACTERS = 'SET_CHARACTERS';
const SHOW_FAVOURITES = 'SHOW_FAVOURITES';
const ApplicationContext = createContext();

const useApplication = () => {
  const [application, setApplication] = useContext(ApplicationContext);
  return { ...application, setApplication };
};

const ApplicationProvider = ({ application = { showFavourites: false }, children }) => {
  const reducer = (state, action) => {
    switch (action.type) {
      case SET_THEME:
        return { ...state, theme: action.value };
      case SET_CHARACTERS:
        return { ...state, characters: action.value };
      case SHOW_FAVOURITES:
        return { ...state, showFavourites: action.value };
      default:
        return state;
    }
  };
  return <ApplicationContext.Provider value={useReducer(reducer, application)}>{children}</ApplicationContext.Provider>;
};

export { useApplication, ApplicationProvider, SET_THEME, SET_CHARACTERS, SHOW_FAVOURITES };
