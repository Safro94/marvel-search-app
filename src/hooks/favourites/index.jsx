import React, { createContext, useReducer, useEffect } from 'react';

import useLocalStorage from '../localStorage';

export const FavouritesContext = createContext();
const ADD_FAVOURITE = 'ADD_FAVOURITE';
const REMOVE_FAVOURITE = 'REMOVE_FAVOURITE';

const reducer = (state, action) => {
  switch (action.type) {
    case ADD_FAVOURITE:
      return [...state, { ...action.value, isFavourite: true }];
    case REMOVE_FAVOURITE:
      return [...state.filter((favourite) => favourite.id !== action.value.id)];
    default:
      return state;
  }
};

const FavouritesProvider = ({ children }) => {
  const { getValueFromLocalStorage, saveValueInLocalStorage } = useLocalStorage();
  const NAME_KEY_STORAGE = 'favourites';
  const [favourites, dispatch] = useReducer(reducer, [], () => getValueFromLocalStorage(NAME_KEY_STORAGE, []));

  useEffect(() => {
    saveValueInLocalStorage(NAME_KEY_STORAGE, favourites);
  }, [favourites, saveValueInLocalStorage]);

  return <FavouritesContext.Provider value={{ favourites, dispatch }}>{children}</FavouritesContext.Provider>;
};
export { FavouritesProvider, ADD_FAVOURITE, REMOVE_FAVOURITE };
