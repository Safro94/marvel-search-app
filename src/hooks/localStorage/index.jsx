const useLocalStorage = () => {
  const hasLocalStorage = () => {
    try {
      return window.localStorage;
    } catch (e) {
      return false;
    }
  };

  const getValueFromLocalStorage = (key, defaultValue) => {
    if (hasLocalStorage()) {
      const valueStorage = localStorage.getItem(key);
      return valueStorage ? JSON.parse(valueStorage) : defaultValue;
    }

    return defaultValue;
  };

  const saveValueInLocalStorage = (key, value) => {
    if (hasLocalStorage()) {
      localStorage.setItem(key, JSON.stringify(value));
    }
  };

  return { hasLocalStorage, getValueFromLocalStorage, saveValueInLocalStorage };
};

export default useLocalStorage;
