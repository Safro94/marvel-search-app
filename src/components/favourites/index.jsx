import React, { useContext } from 'react';

import { useApplication, SET_CHARACTERS, SHOW_FAVOURITES } from '../../hooks/application';
import { FavouritesContext } from '../../hooks/favourites';

import Star from '../shared/star';

export default () => {
  const { setApplication, showFavourites } = useApplication();
  const { favourites } = useContext(FavouritesContext);

  const handleClick = () => {
    const characters = !showFavourites ? favourites : [];
    setApplication({ type: SET_CHARACTERS, value: characters });
    setApplication({ type: SHOW_FAVOURITES, value: !showFavourites });
  };

  return <Star handleClick={handleClick} showFullStar={showFavourites} data-testid="star" />;
};
