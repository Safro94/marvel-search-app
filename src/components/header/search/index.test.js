/* eslint-disable import/first */
import React from 'react';
import { render, wait, waitForElement, fireEvent } from '@testing-library/react';

import Search from './';

jest.mock('../../../services/service', () => {
  return {
    GetCharacters: jest.fn().mockImplementation(() => {
      return [
        { id: 1, title: 'title' },
        { id: 2, title: 'title2' },
      ];
    }),
  };
});
import Service from '../../../services/service';

jest.mock('./style', () => {
  return {
    Input: 'input',
  };
});

jest.mock('../../../hooks/application', () => ({
  useApplication: jest.fn(() => ({
    setApplication: jest.fn(),
  })),
}));

import { FavouritesContext } from '../../../hooks/favourites';

describe('<Search /> ', () => {
  const history = [];
  beforeEach(() => {
    history.length = 0;
    jest.clearAllMocks();
  });

  const favourites = [{ id: 1 }];

  const TestComponent = () => (
    <FavouritesContext.Provider value={{ favourites }}>
      <Search.WrappedComponent history={history} />
    </FavouritesContext.Provider>
  );

  it('It should render OK', async () => {
    // Act
    const { container } = render(<TestComponent />);
    await wait();

    // Assert
    expect(container).toMatchSnapshot();
  });

  it('When the input onBlur is called, it should call make a history push', async () => {
    //Arrange
    const url = '/comic/111';

    // Act
    const { getByTestId } = render(<TestComponent />);
    const input = await waitForElement(() => getByTestId('input'));
    fireEvent.blur(input, { target: { value: 'https://www.marvel.com/title/test/111/name' } });

    // Assert
    expect(history[0]).toBe(url);
  });

  it('When the input onBlur is called, it should call GetCharacters', async () => {
    // Act
    const { getByTestId } = render(<TestComponent />);
    const input = await waitForElement(() => getByTestId('input'));
    fireEvent.blur(input, { target: { value: 'spider' } });

    // Assert
    expect(Service.GetCharacters).toBeCalledTimes(1);
  });
});
