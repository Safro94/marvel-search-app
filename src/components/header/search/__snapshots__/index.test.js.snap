// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search />  It should render OK 1`] = `
<div>
  <input
    data-testid="input"
    placeholder="Buscar"
  />
</div>
`;
